lib/theme_check/checks/pagination_size.rb

```diff
@@ -0,0 +1,61 @@
+# frozen_string_literal: true
+module ThemeCheck
+  class PaginationSize < LiquidCheck
+    severity :suggestion
+    categories :performance
+    doc docs_url(__FILE__)
+
+    attr_reader :max_size
+
+    def initialize(max_size: 50)
+      @max_size = max_size
+    end
+
+    def on_document(_node)
+      @paginations = {}
+      @schema_settings = {}
+    end
+
+    def on_paginate(node)
+      size = node.value.page_size
+      unless @paginations.key?(size)
+        @paginations[size] = []
+      end
+      @paginations[size].push(node)
+    end
+
+    def on_schema(node)
+      schema = JSON.parse(node.value.nodelist.join)
+
+      if (settings = schema["settings"])
+        @schema_settings = settings
+      end
+    rescue JSON::ParserError
+      # Ignored, handled in ValidSchema.
+    end
+
+    def after_document(_node)
+      @paginations.each_pair do |size, nodes|
+        numerical_size = if size.is_a?(Numeric)
+          size
+        else
+          get_setting_default_value(size.lookups.last)
+        end
+        if numerical_size > @max_size
+          nodes.each { |node| add_offense("Use a smaller pagination size", node: node) }
+        end
+      end
+    end
+
+    private
+
+    def get_setting_default_value(setting_id)
+      setting = @schema_settings.select { |s| s['id'] == setting_id }
+      unless setting.empty?
+        return setting.first['default']
```

Comment:
also, `.select` returns an array of all matching settings by 'id' - suppose the user sets the settings twice by mistake. wouldn't we want the second setting to overwrite the first?  rather than vice versa?  in that case we would want:
```
return setting.last['default']
```

Reply:
@nresh I can make that update. I was assuming that a check in `valid_schema.rb` would prevent duplicate settings, so that we're not having to account for that.